### ISSUE TITLE
Maybe::eval() with lazy default?

### ISSUE THREAD

**Issue by someniatko** [marcosh/lamphpda#39]
`Maybe::eval()` requires me to pass the default value for the `nothing` case immediately. Instead, I would like to evaluate it lazily, only when I really reach the `nothing` state. For that I would like to pass a callback accepting no arguments. Is it a feasible addition?

**Comment by someniatko** [comment]
Thank you for a really fast answer, I didn't expect that :). I am not sure if I its okay to use Either, if I don't want to specify any additional information for the "nothing" state. Either is fine e.g. for failable actions, where you want to know exact failure reason, but in my case semantically Maybe is more correct, therefore I'd like to use it.

**Comment by drupol** [comment]
Ok !

How about just passing a callable to `Maybe` then ?

**Comment by someniatko** [comment]
Idk how would it help?

Let's consider a case like the following:

An HTTP controller
```php
/** @var Maybe<Entity> */
$maybeEntity = $repository->findById('some-id');

return $maybeEntity->eval(
    $this->buildNotFoundResponse(),
    fn (Entity $e) => $this->buildResponseWithEntity($e),
);
```

and I would like not to create `$this->buildNotFoundResponse()` until I really need to. It may be something else, much more computationally heavy.

**Comment by drupol** [comment]
How about something like:

```php
/** @var Maybe<Entity> */
$maybeEntity = $repository->findById('some-id');

return $maybeEntity->eval(
    static fn () => $this->buildNotFoundResponse(),
    fn (Entity $e) => $this->buildResponseWithEntity($e),
);
```

**Comment by drupol** [comment]
Also,

If you're using Doctrine repositories, maybe this might be interesting for you: https://github.com/loophp/repository-monadic-helper/

**Comment by someniatko** [comment]
Yes, that's exactly how I would like to use this lib. However, in the current implementation my controller will just return a callable instead of a response in the `nothing` case.

**Comment by drupol** [comment]
Also,

This:

```php
/** @var Maybe<Entity> */
$maybeEntity = $repository->findById('some-id');

return $maybeEntity->eval(
    static fn () => $this->buildNotFoundResponse(),
    fn (Entity $e) => static fn () => $this->buildResponseWithEntity($e),
)();
```

and

```php
/** @var Maybe<Entity> */
$maybeEntity = $repository->findById('some-id');

return $maybeEntity->eval(
    $this->buildNotFoundResponse(),
    fn (Entity $e) => $this->buildResponseWithEntity($e),
);
```

is exactly the same. So I don't see your point.

Maybe @marcosh ?

**Comment by someniatko** [comment]
You are right that it's the same in terms of the final result (at least if we assume that the callables do not cause side-effects, but we cannot guarantee it, we're not in Haskell y'know). However, it's not the same non-functionally. The second variant **always** executes `buildNotFoundResponse()`. My point was I don't want to call `buildNotFoundResponse()` function altogether. Imagine it is taking 1 second to execute. I don't want to waste this second if I am not in `nothing` case.

**Comment by drupol** [comment]
allright, fully got your point now. I'm now running out of ideas :=)

Let's wait for Marco to shime in, he most probably has a better idea.

**Comment by marcosh** [comment]
@someniatko one thing I notice in your example is that your are a bit too eager to use `eval`. `eval` should be the last thing you call, after you performed all your computations.

What I would suggest instead is to first `map` over `$maybeEntity` to compute your final value and only then `eval`.

This does not solve your issue, but it could still be an improvement.

In general, if you want your data to be lazy, you should model them like so. In the context of this library, laziness is modelled using `IO`, which is equivalent to the approach @drupol proposed, but could make it look a bit nicer

**Comment by someniatko** [comment]
Okay, instead I could first `map` it for the "just" case and then somehow "unwrap" the Maybe, optionally providing a default value, preferably lazily. Is there a method to "unwrap" Maybe? 

By "unwrapping" i mean the following Haskell-like scheme:
`Maybe<A> -> (() -> A) -> A`

**Comment by someniatko** [comment]
BTW, I am not sure in Haskell you need to use `IO` for the reasons outlined here, because Haskell itself is a lazilly evaluated language. However, PHP isn't! Therefore IMO we need to adapt functional APIs accordingly due to this fact.

**Comment by marcosh** [comment]
Haskell is lazy, yes. But it works this way even in other strict languages like [Purescript](https://github.com/purescript/purescript-maybe/blob/c6f98ac1088766287106c5d9c8e30e7648d36786/src/Data/Maybe.purs#L237-L237).

 [Typescript](https://github.com/gcanti/fp-ts/blob/820b78d0d9aa81330cd8ce117289c7b97b6f9d14/src/Option.ts#L840) does something a little bit different, requesting an explicitly lazy input.

I actually take back my `IO` suggestion, because that was taking advantage of the internal encoding of `IO`, which I would like to keep private.

@drupol What we could do is add a method `withLazyDefault :: (() -> a) -> Maybe a -> a`. @someniatko meanwhile, you could implement it yourself as `$maybe->map(fn ($a) => fn () => $a)->withDefault($f)()` (hopefully I got it right without trying to execute it...)

**Comment by drupol** [comment]
That would definitely be doable.

**Comment by someniatko** [comment]
Yeah, but following the Purescript link, the very next method is `maybe'` with lazy default :)

**Comment by someniatko** [comment]
it seems now it's solved by #41
